src/ngraph/op/fused/depth_to_space.cpp

```diff
@@ -46,59 +47,101 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
-                 "The provided tensor shape: ",
-                 data_shape,
-                 " is not supported.");
-
-    // Assume NCHW data layout
-    if (data_shape.size() == 4)
-    {
-        n = data_shape.at(0);
-        c = data_shape.at(1);
-        h = data_shape.at(2);
-        w = data_shape.at(3);
-    }
-    // Without batch.
-    else if (data_shape.size() == 3)
+    if (data_shape.size() == 3)
     {
-        c = data_shape.at(0);
-        h = data_shape.at(1);
-        w = data_shape.at(2);
+        // Insert batch axis
+        data_shape.insert(data_shape.begin(), 1);
+        data = builder::reshape(data, data_shape);
     }
-
-    NGRAPH_CHECK((c % (m_blocksize * m_blocksize) == 0 && m_blocksize > 0),
-                 "SpaceToDepth: The depth axis size must be a multiple of ",
-                 "squared block_size attribute value.");
+    const size_t n_dim = data_shape.at(0);
+    const size_t c_dim = data_shape.at(1);
+    const size_t spatial_dim_index = 2;
+    const size_t spatial_dims = data_shape.size() - spatial_dim_index;
+    const auto c_dim_divider = static_cast<int>(std::pow(m_blocksize, spatial_dims));
+
+    NODE_VALIDATION_CHECK(this,
+                          m_blocksize > 0 && c_dim % c_dim_divider == 0,
+                          "DepthToSpace: The input data's 'channels' axis size: ",
+                          c_dim,
+                          " must be a equivalent to ",
+                          "'block_size'^'spatial_dims': ",
+                          c_dim_divider);
 
     auto bs = static_cast<size_t>(m_blocksize);
-    size_t c_flat = c / (bs * bs);
+    size_t c_flat = c_dim / c_dim_divider;
 
     // First we have to disperse the data from depth channel, then rearrange them
     // so as appropriate chunks of data where close to their destination place.
     // Finally squeeze data from respective dimensions.
     shared_ptr<Node> flat_node;
+    Shape dispersed_shape{n_dim};
+    for (int i = 0; i < spatial_dims; ++i)
+    {
+        dispersed_shape.push_back(bs);
+    }
+    for (int i = 0; i < spatial_dims; ++i)
+    {
+        dispersed_shape.push_back(data_shape.at(spatial_dim_index + i));
+    }
+    vector<size_t> axes_order{0};
     switch (m_mode)
     {
+    // x' = reshape(data, [N, C / (block_size ^ K), block_size, block_size, ..., block_size, D1, D2,
+    // ..., DK])
+    // x'' = transpose(x', [0,  1,  K + 2, 2, K + 3, 3, K + 4, 4, ..., K + (K + 1), K + 1])
+    // y = reshape(x'', [N, C / (block_size ^ K), D1 * block_size, D2 * block_size, D3 * block_size,
+    // ..., DK * block_size])
     case DepthToSpaceMode::DEPTH_FIRST:
     {
-        flat_node = builder::reshape(data, Shape{n, c_flat, bs, bs, h, w});
-        flat_node = builder::reorder_axes(flat_node, {0, 1, 4, 2, 5, 3});
+        dispersed_shape.insert(dispersed_shape.begin() + 1, c_flat);
+        flat_node = builder::reshape(data, dispersed_shape);
```

Comment:
Will it produce V1::Reshape? 
Due to we have to decompose this op in conversion layer the best way is to reuse operation decomposition method for both DepthToSpace and SpaceToDepth.

Reply:
As agreed, I'll add producing `Reshape:v1` in separate PR